File: os/axvisor/web-ui/src/domain/files.ts
//! The file-transfer state machine: one implementation, several consumers.
//!
//! The file panel drives it with a drop target; the creation form drives it to
//! put a file where a config names it. Neither panel reaches into the other —
//! the composition root (`shell/App.tsx`) builds this service from the `files`
//! resource's declared operations and hands it down, so the transfer protocol
//! lives in exactly one place.
//!
//! The operations are the ones the hypervisor already declares, one per step:
//!
//! - **open** starts a session from a client-chosen id, so sending the same file
//!   twice resumes the first attempt instead of staging the bytes twice;
//! - **send** carries one chunk per request, so what a transfer holds in memory
//!   is bounded by the chunk size rather than by the file;
//! - **resume** (`HEAD`) is how an interrupted transfer asks where it stopped.
//!   The answer is the length of the staged file on disk, which is the only
//!   offset a client may continue from;
//! - **place** moves the staged bytes to their final name, and is the step that
//!   can be refused because that name is taken;
//! - **drop** forgets a session, and is refused once the file is placed: those
//!   bytes are what a config may already reference;
//! - **mkdir** is the one directory the interface creates, because a transfer
//!   target has to exist before bytes can go into it.
//!
//! Nothing here spells a request path: every URL comes from the injected link.

import { useSyncExternalStore } from 'react'
import type { ApiClient } from '@/api/client'
import {
  ApiError,
  describeError,
  type FileSession,
  type FileState,
  type FilesCapability,
  type FilesInfo,
  type FilesSnapshot,
  type PanelLink,
  type Transfer,
} from '@/api/types'

/**
 * Bytes per chunk request.
 *
 * The hypervisor refuses a chunk larger than its own limit and does not declare
 * that limit in the manifest, so this stays at or below it; the file is never
 * read whole, which is what keeps a large transfer off the heap.
 */
export const CHUNK_BYTES = 1024 * 1024

/** Longest session id the control plane accepts. */
const SESSION_ID_MAX = 64

/** Response header that reports how many bytes are on disk. */
const UPLOAD_OFFSET = 'Upload-Offset'

/** One rendered row, from the backend's listing or from a transfer in flight. */
export interface Row {
  id: string
  name: string
  directory: string
  written: number
  total: number
  state: FileState
  detail: string | null
  /** Present while this row is a transfer this service is driving. */
  transfer: Transfer | null
}

const EMPTY_SNAPSHOT: FilesSnapshot = { sessions: [], transfers: [] }

/**
 * The transfers and the session listing, as one observable value.
 *
 * A consumer subscribes and reads: the service is the store, so a transfer
 * started anywhere (the file panel, a creation form) shows up in every consumer
 * without either of them knowing about the other.
 */
export class FilesService implements FilesCapability {
  private sessions: FileSession[] = []
  private transfers: Transfer[] = []
  private snapshot: FilesSnapshot = EMPTY_SNAPSHOT
  private readonly listeners = new Set<() => void>()
  private readonly aborts = new Map<string, AbortController>()

  constructor(
    private readonly api: ApiClient,
    private readonly link: PanelLink,
  ) {}

  subscribe = (listener: () => void): (() => void) => {
    this.listeners.add(listener)
    return () => {
      this.listeners.delete(listener)
    }
  }

  /** Stable between notifications, which is what `useSyncExternalStore` needs. */
  getSnapshot = (): FilesSnapshot => this.snapshot

  private publish(): void {
    this.snapshot = { sessions: this.sessions, transfers: this.transfers }
    for (const listener of this.listeners) listener()
  }

  private track(row: Transfer): void {
    const known = this.transfers.find((candidate) => candidate.id === row.id)
    this.transfers = known
      ? this.transfers.map((candidate) =>
          candidate.id === row.id ? { ...candidate, ...row } : candidate,
        )
      : [...this.transfers, row]
    this.publish()
  }

  private update(id: string, patch: Partial<Transfer>): void {
    this.transfers = this.transfers.map((row) => (row.id === id ? { ...row, ...patch } : row))
    this.publish()
  }

  /** Reads the backend's session listing. */
  async refresh(): Promise<void> {
    const info = await this.api.get<FilesInfo>(this.link.url('list'))
    this.sessions = info.files
    this.publish()
  }

  /**
   * Sends one file into `directory`, finally named `name`.
   *
   * `name` is the final name rather than always the file's own: a creation form
   * transfers to the exact path its config names, which may differ from the name
   * the operator picked the file under. A refusal is recorded on the transfer
   * instead of thrown, because a drop and a form submission both start one
   * without waiting for it.
   */
  async upload(file: File, directory: string, name: string = file.name): Promise<void> {
    const id = sessionId(file)
    this.aborts.set(id, new AbortController())
    this.track({
      id,
      name,
      directory,
      written: 0,
      total: file.size,
      phase: 'opening',
      detail: null,
      file,
    })
    try {
      // `open` is idempotent: sending the same file again answers with the
      // session it already has and the bytes already on disk, so a fresh send
      // and a restart share this path.
      const session = await this.api.post<FileSession>(this.link.url('open'), {
        id,
        directory,
        total: file.size,
      })
      await this.sendFrom(file, id, session.written, name)
    } catch (e: unknown) {
      if (e instanceof DOMException && e.name === 'AbortError') {
        this.update(id, { phase: 'failed', detail: '已取消' })
        return
      }
      this.update(id, { phase: 'failed', detail: describeError(e) })
    } finally {
      this.aborts.delete(id)
    }
  }

  /** Continues a transfer whose bytes this service still holds. */
  async resume(id: string): Promise<void> {
    const transfer = this.transfers.find((row) => row.id === id)
    const file = transfer?.file ?? null
    if (transfer === undefined || file === null) return
    this.aborts.set(id, new AbortController())
    try {
      // The offset comes from the disk, through the plane that owns it: a local
      // counter could be ahead of what survived the interruption.
      const offset = offsetOf(await this.api.head(this.link.url('resume', { id })))
      await this.sendFrom(file, id, offset, transfer.name)
    } catch (e: unknown) {
      this.update(id, { phase: 'failed', detail: describeError(e) })
    } finally {
      this.aborts.delete(id)
    }
  }

  /** Sends one file from `offset` to the end, then places it as `name`. */
  private async sendFrom(file: File, id: string, offset: number, name: string): Promise<void> {
    const signal = this.aborts.get(id)?.signal
    let at = offset
    this.update(id, { phase: 'sending', written: at })
    while (at < file.size) {
      const end = Math.min(at + CHUNK_BYTES, file.size)
      const headers = await this.api.patch(
        this.link.url('send', { id }),
        file.slice(at, end),
        {
          'Content-Type': 'application/octet-stream',
          'Content-Range': `bytes ${at}-${end - 1}/${file.size}`,
        },
        signal,
      )
      at = offsetOf(headers)
      this.update(id, { written: at })
    }
    await this.place(id, name)
  }

  /**
   * Moves finished bytes to their final name.
   *
   * A taken name is the one refusal that is about the request rather than the
   * transfer: the staged bytes are complete and stay complete, so the row asks
   * for another name instead of offering a resume. The refusal is recorded and
   * not thrown, so the caller that started a transfer does not have to catch it.
   */
  async place(id: string, name: string): Promise<boolean> {
    this.update(id, { phase: 'placing', detail: null })
    try {
      const placed = await this.api.post<FileSession>(this.link.url('place', { id }), { name })
      this.update(id, { phase: 'placed', name, written: placed.written, detail: null, file: null })
      await this.refresh()
      return true
    } catch (e: unknown) {
      const taken = e instanceof ApiError && e.status === 409
      this.update(id, { phase: taken ? 'needs-name' : 'failed', detail: describeError(e) })
      return false
    }
  }

  /** Forgets a session and the bytes it staged. */
  async drop(id: string): Promise<void> {
    // A transfer still running has to stop before its session goes away, or the
    // next chunk would arrive for a session that is gone.
    this.aborts.get(id)?.abort()
    await this.api.del(this.link.url('drop', { id }))
    this.transfers = this.transfers.filter((row) => row.id !== id)
    this.publish()
    await this.refresh()
  }

  /** Creates one directory level, returning the path it created. */
  async mkdir(parent: string, name: string): Promise<string> {
    const made = await this.api.post<{ path: string }>(this.link.url('mkdir'), {
      parent: parent.trim() || '/',
      name: name.trim(),
    })
    return made.path
  }
}

const NO_SUBSCRIPTION = (): (() => void) => () => {}
const NO_SNAPSHOT = (): FilesSnapshot => EMPTY_SNAPSHOT

/** Observes a service, or an empty value when this build has no transfer panel. */
export function useFiles(service: FilesCapability | null): FilesSnapshot {
  return useSyncExternalStore(
    service?.subscribe ?? NO_SUBSCRIPTION,
    service?.getSnapshot ?? NO_SNAPSHOT,
  )
}

/**
 * The rows to show: the backend's listing, plus the transfers in flight.
 *
 * A transfer in flight is not in the listing at all (the control plane does not
 * publish `uploading`), and a transfer that just ended is in both. So a local
 * row wins while it is still moving, and the listing wins once it has settled:
 * that way a row never appears twice and never shows a stale phase.
 */
export function mergeRows(sessions: FileSession[], transfers: Transfer[]): Row[] {
  const byId = new Map<string, Row>()
  for (const session of sessions) {
    byId.set(session.id, {
      id: session.id,
      name: session.name ?? '',
      directory: session.directory,
      written: session.written,
      total: session.total,
      state: session.state,
      detail: session.detail,
      transfer: null,
    })
  }
  for (const transfer of transfers) {
    const moving = isMoving(transfer)
    if (!moving && byId.has(transfer.id)) continue
    byId.set(transfer.id, {
      id: transfer.id,
      name: transfer.name,
      directory: transfer.directory,
      written: transfer.written,
      total: transfer.total,
      state: stateOf(transfer),
      detail: transfer.detail,
      transfer,
    })
  }
  return [...byId.values()]
}

/** Whether a transfer is still in the part of its life the listing cannot show. */
export function isMoving(transfer: Transfer): boolean {
  return (
    transfer.phase === 'opening' ||
    transfer.phase === 'sending' ||
    transfer.phase === 'placing' ||
    transfer.phase === 'needs-name'
  )
}

/** The reported state a locally tracked transfer corresponds to. */
export function stateOf(transfer: Transfer): FileState {
  switch (transfer.phase) {
    case 'opening':
    case 'sending':
      return 'uploading'
    case 'placing':
      return 'placing'
    case 'placed':
      return 'placed'
    case 'needs-name':
      return 'uploaded'
    case 'failed':
      return 'failed'
  }
}

/**
 * The id one file is staged under.
 *
 * Stable for one file, so sending it again resumes the first attempt, and
 * distinct for a different file that happens to share its name and length: the
 * modification stamp is part of it. The control plane accepts ascii
 * alphanumerics plus `-_.+` as a single path component.
 */
export function sessionId(file: File): string {
  const stem = file.name
    .replace(/[^A-Za-z0-9-_.+]/g, '-')
    // Two dots anywhere are refused by the control plane, and a leading one too,
    // because the id becomes one path component of the staging file. A single
    // dot is kept so the id still reads like the file it belongs to.
    .replace(/\.{2,}/g, '-')
    .replace(/^\.+/, '-')
    .slice(0, 32)
  return `${stem}-${file.size}-${file.lastModified}`.slice(0, SESSION_ID_MAX)
}

/** Reads the reported offset, which is the only position a client may trust. */
export function offsetOf(headers: Headers): number {
  const raw = headers.get(UPLOAD_OFFSET)
  const offset = raw === null ? Number.NaN : Number(raw)
  if (!Number.isFinite(offset)) {
    throw new Error('后端没有报告 Upload-Offset，无法确认已落盘长度')
  }
  return offset
}

File: os/axvisor/src/control/domain/files.rs
//! Files on their way into a guest filesystem.
//!
//! A file becomes usable by a guest config only once it sits at the path the
//! config names, so this module has one job: move bytes there without ever
//! exposing a half-transferred object. It is a *staging* store, not the guest
//! filesystem itself: bytes arrive in chunks into a namespace of our own, and a
//! separate placement step puts the finished file where the config will look
//! for it.
//!
//! # Why the bytes are staged at all
//!
//! The target may be a file a running guest already holds open (its disk, its
//! kernel), and the running guest is the owner of that file, not this plane.
//! Writing into it chunk by chunk would let a guest read a half-written disk.
//! Staging keeps the target untouched until every byte has arrived, and
//! [`place`] then refuses a target that already exists rather than replacing
//! something another guest may be reading. That refusal is also what keeps this
//! plane off a live guest's disk: such a path exists by definition, so a file
//! another guest is using can never be the target of a placement.
//!
//! # Where the truth lives
//!
//! The session table below is memory; the *offset* is not. Every entry point
//! that has to know how many bytes are already on disk asks the file itself
//! (`Metadata::len`), so a client that reconnects — or a control plane that
//! restarted — resumes from what is really there instead of from a counter that
//! promises more than the disk holds. The table therefore only carries what the
//! disk cannot say: which directory a session belongs to, how long it promised
//! to be, and where it has got to in the walk below.
//!
//! # States
//!
//! `uploading` are the bytes still arriving, `uploaded` the bytes that are
//! complete, `placing` the moment the target is being taken, `placed` the file
//! in its final path, and `failed` an attempt that stopped with a readable
//! reason. A full disk is not terminal: the bytes already written stay valid, so
//! a `send` after the operator frees space continues from the same offset.

use alloc::collections::BTreeMap;
use alloc::format;
use alloc::string::{String, ToString};
use alloc::vec::Vec;
use std::sync::LazyLock;

use ax_std::StdError;
use ax_std::fs::OpenOptions;
use ax_std::io::Error as IoError;
use ax_std::io::Write;
use ax_std::os::arceos::sync::NoPreemptMutex;

use super::pool;

/// The largest chunk a client may send in one request.
///
/// It is the request-body limit as well: the transport layer reads one chunk
/// per request, so this number is also the memory a single upload can occupy.
/// One mebibyte keeps a chunk small enough that serving it does not hold the
/// control-plane thread for long, and large enough that the per-request cost
/// stays in the noise.
pub const CHUNK_LIMIT: usize = 1024 * 1024;

/// The largest file this plane will accept, declared or streamed.
///
/// A guest disk image is the biggest thing that travels through here; 256 MiB
/// covers the images this workspace builds and keeps a runaway client from
/// filling the hypervisor's own filesystem.
pub const FILE_LIMIT: usize = 256 * 1024 * 1024;

/// Name of the staging namespace inside the target directory.
///
/// It starts with a dot, so it is not a name a config or a pool file can take
/// (the pool rejects dot-files), and it keeps staged objects out of the listing
/// a person reads while looking for something to boot.
pub const STAGING_DIR: &str = ".files";

/// Longest accepted session identifier, before it becomes a path component.
const SESSION_ID_MAX: usize = 64;

/// One staged object, as the API reports it.
pub struct SessionView {
    pub id: String,
    pub directory: String,
    /// Final name, known once `place` has been asked for it.
    pub name: Option<String>,
    /// Final path, known once the file is `placed`.
    pub path: Option<String>,
    pub total: usize,
    /// Bytes actually on disk, which is the only offset this plane trusts.
    pub written: usize,
    pub state: &'static str,
    pub detail: Option<String>,
}

/// Why an operation could not be carried out.
///
/// The variants name the *kind* of failure, not the status code: mapping a
/// failure to HTTP belongs to the transport layer, which is the layer that
/// speaks HTTP.
pub enum FileError {
    /// The session id is not a name this plane will turn into a path.
    InvalidId(String),
    /// The target directory is not an absolute path inside the guest filesystem.
    InvalidDirectory(String),
    /// The target directory does not exist.
    ///
    /// Nothing creates it on the way in: the operator picks a folder, and a
    /// folder that is not there is a mistake or a step that was skipped. The
    /// explicit operation for making one is [`make_directory`].
    NoSuchDirectory(String),
    /// A directory with that name is already there.
    Exists(String),
    /// The final name is not a plain file name.
    InvalidName(String),
    /// The declared or streamed size exceeds [`FILE_LIMIT`].
    TooLarge { total: usize, limit: usize },
    /// No session by that id.
    Unknown(String),
    /// The request does not fit the state the session is in. `offset` is where
    /// the client should continue from.
    Conflict { reason: String, offset: usize },
    /// The guest filesystem refused to write.
    Unwritable(String),
    /// The guest filesystem is out of space. Recoverable: the bytes already
    /// written stay valid, so the same session can continue after cleanup.
    StorageFull(String),
}

/// Where one session stands.
#[derive(Clone)]
enum State {
    Uploading,
    Uploaded,
    Placing,
    Placed,
    Failed(String),
}

impl State {
    fn name(&self) -> &'static str {
        match self {
            State::Uploading => "uploading",
            State::Uploaded => "uploaded",
            State::Placing => "placing",
            State::Placed => "placed",
            State::Failed(_) => "failed",
        }
    }
}

struct Session {
    directory: String,
    total: usize,
    name: Option<String>,
    path: Option<String>,
    state: State,
}

static SESSIONS: LazyLock<NoPreemptMutex<BTreeMap<String, Session>>> =
    LazyLock::new(|| NoPreemptMutex::new(BTreeMap::new()));

/// Opens a session, or picks up the one an earlier attempt left behind.
///
/// `total` is what the client promises the file will be. A session that already
/// exists keeps the promise it was opened with, and the offset that comes back
/// is always read from the disk: dragging the same file twice therefore resumes
/// instead of starting over, and never truncates bytes that already arrived. To
/// start over deliberately, `drop` the session first.
pub fn open(id: &str, directory: &str, total: usize) -> Result<SessionView, FileError> {
    let id = checked_id(id)?;
    let directory = checked_directory(directory)?;
    if total > FILE_LIMIT {
        return Err(FileError::TooLarge {
            total,
            limit: FILE_LIMIT,
        });
    }

    if ax_std::fs::read_dir(&directory).is_err() {
        return Err(FileError::NoSuchDirectory(directory));
    }
    // The staging namespace is this plane's own, not the operator's tree, so it
    // is the one directory the transfer may create on the way in.
    let staging = staging_dir(&directory);
    pool::ensure_directory(&staging).map_err(FileError::Unwritable)?;
    let path = staging_path(&directory, id);
    // `create` without `truncate`: a resumed upload must find its bytes, and a
    // second `open` for the same id must not throw them away.
    OpenOptions::new()
        .create(true)
        .write(true)
        .open(&path)
        .map_err(|error| map_std_error(&error))?;

    let written = disk_len(&path);
    let mut sessions = SESSIONS.lock();
    let session = sessions.entry(id.to_string()).or_insert_with(|| Session {
        directory: directory.clone(),
        total,
        name: None,
        path: None,
        state: State::Uploading,
    });
    session.state = if written >= session.total {
        State::Uploaded
    } else {
        State::Uploading
    };
    Ok(view(id, session, written))
}

/// Creates one directory level inside the guest filesystem.
///
/// This is the explicit "new folder" of the interface, one level at a time for
/// the same reason the config pool writes one level at a time: the guest
/// filesystem cannot create a chain of directories. The parent has to exist, and
/// a name that is already taken is a conflict rather than a silent suffix — a
/// client that wants "name (2)" asks for that name explicitly, so the interface
/// chooses it and the plane never invents one.
pub fn make_directory(parent: &str, name: &str) -> Result<String, FileError> {
    let parent = checked_directory(parent)?;
    let name = checked_directory_name(name)?;
    let path = format!("{parent}/{name}");
    if ax_std::fs::read_dir(&parent).is_err() {
        return Err(FileError::NoSuchDirectory(parent));
    }
    if ax_std::fs::read_dir(&path).is_ok() {
        return Err(FileError::Exists(path));
    }
    pool::ensure_directory(&path).map_err(FileError::Unwritable)?;
    Ok(path)
}

/// How many bytes are already on disk for one session.
///
/// Placed sessions report the finished length instead: their bytes are no
/// longer staged, and a client that asks again should not be told the file it
/// finished is empty.
pub fn resume(id: &str) -> Result<usize, FileError> {
    let id = checked_id(id)?;
    let (directory, state) = {
        let sessions = SESSIONS.lock();
        let session = sessions.get(id).ok_or_else(|| unknown(id))?;
        (session.directory.clone(), session.state.clone())
    };
    let written = disk_len(&staging_path(&directory, id));
    match state {
        State::Placed => Err(FileError::Conflict {
            reason: format!("{id} is placed"),
            offset: written,
        }),
        _ => Ok(written),
    }
}

/// Appends one chunk and reports the new offset.
///
/// The chunk's own start is compared against the length on disk: a client that
/// reconnects mid-flight has a stale idea of where it stopped, and answering
/// with [`FileError::Conflict`] plus the real offset is what makes the retry
/// land in the right place.
pub fn send(id: &str, start: usize, bytes: &[u8]) -> Result<usize, FileError> {
    let id = checked_id(id)?;
    let (directory, total) = {
        let sessions = SESSIONS.lock();
        let session = sessions.get(id).ok_or_else(|| unknown(id))?;
        (session.directory.clone(), session.total)
    };

    let path = staging_path(&directory, id);
    let written = disk_len(&path);
    if start != written {
        return Err(FileError::Conflict {
            reason: format!("chunk starts at {start} but {written} bytes are on disk"),
            offset: written,
        });
    }
    let end = written + bytes.len();
    if end > FILE_LIMIT {
        return Err(FileError::TooLarge {
            total: end,
            limit: FILE_LIMIT,
        });
    }
    if end > total {
        return Err(FileError::Conflict {
            reason: format!("{end} bytes exceed the declared {total}"),
            offset: written,
        });
    }

    // `append` and `write` go together: append decides *where* the bytes go,
    // write decides that the handle may write at all.
    let mut file = OpenOptions::new()
        .append(true)
        .write(true)
        .open(&path)
        .map_err(|error| map_std_error(&error))?;
    if let Err(error) = file.write_all(bytes) {
        record_failure(id, &error.to_string());
        return Err(map_io_error(&error));
    }

    let mut sessions = SESSIONS.lock();
    if let Some(session) = sessions.get_mut(id) {
        session.state = if end >= session.total {
            State::Uploaded
        } else {
            State::Uploading
        };
    }
    Ok(end)
}

/// Moves a finished object to the path a guest config will name.
///
/// The target must not exist. That is not a convenience: a config names its
/// kernel and its disks, and replacing a file some guest already boots from
/// would change that guest's contents behind its back, so a conflict is the only
/// honest answer and the client decides what to do about it.
pub fn place(id: &str, name: &str) -> Result<SessionView, FileError> {
    let id = checked_id(id)?;
    let name = checked_name(name)?;
    let (directory, total, state) = {
        let sessions = SESSIONS.lock();
        let session = sessions.get(id).ok_or_else(|| unknown(id))?;
        (
            session.directory.clone(),
            session.total,
            session.state.clone(),
        )
    };

    let written = match state {
        State::Uploaded => disk_len(&staging_path(&directory, id)),
        State::Placed => {
            return Err(FileError::Conflict {
                reason: format!("{id} is already placed"),
                offset: total,
            });
        }
        State::Placing => {
            return Err(FileError::Conflict {
                reason: format!("{id} is being placed"),
                offset: total,
            });
        }
        State::Uploading => {
            let written = disk_len(&staging_path(&directory, id));
            return Err(FileError::Conflict {
                reason: format!("{id} is not complete: {written} of {total} bytes"),
                offset: written,
            });
        }
        State::Failed(ref detail) => {
            return Err(FileError::Conflict {
                reason: format!("{id} failed earlier: {detail}"),
                offset: disk_len(&staging_path(&directory, id)),
            });
        }
    };
    if written != total {
        return Err(FileError::Conflict {
            reason: format!("{written} bytes on disk, {total} declared"),
            offset: written,
        });
    }

    let from = staging_path(&directory, id);
    let to = format!("{directory}/{name}");
    if ax_std::fs::metadata(&to).is_ok() {
        return Err(FileError::Conflict {
            reason: format!("{to} already exists"),
            offset: written,
        });
    }

    {
        let mut sessions = SESSIONS.lock();
        if let Some(session) = sessions.get_mut(id) {
            session.state = State::Placing;
            session.name = Some(name);
        }
    }
    if let Err(error) = ax_std::fs::rename(&from, &to) {
        record_failure(id, &error.to_string());
        return Err(map_std_error(&error));
    }

    let mut sessions = SESSIONS.lock();
    let session = sessions.get_mut(id).ok_or_else(|| unknown(id))?;
    session.state = State::Placed;
    session.path = Some(to);
    Ok(view(id, session, total))
}

/// Forgets a session and removes the bytes it staged.
///
/// A placed session is not dropped: its bytes are no longer staged, they are
/// the file a config points at, and deleting them here would be a deletion the
/// operator never asked for.
pub fn drop(id: &str) -> Result<(), FileError> {
    let id = checked_id(id)?;
    let (directory, state) = {
        let sessions = SESSIONS.lock();
        let session = sessions.get(id).ok_or_else(|| unknown(id))?;
        (session.directory.clone(), session.state.clone())
    };
    if let State::Placed = state {
        return Err(FileError::Conflict {
            reason: format!("{id} is placed; its bytes are a file a config may name"),
            offset: 0,
        });
    }

    let path = staging_path(&directory, id);
    // Not being there is the outcome the caller asked for, so only a genuine
    // failure to remove something that exists is an error.
    if let Err(error) = ax_std::fs::remove_file(&path) {
        if ax_std::fs::metadata(&path).is_ok() {
            return Err(map_std_error(&error));
        }
    }
    SESSIONS.lock().remove(id);
    Ok(())
}

/// Every session a client can still act on, by id.
///
/// Sessions whose bytes are still arriving are left out on purpose: a listing is
/// how an operator picks something to boot, and an object that is not whole yet
/// must not appear there. The client that is uploading already knows its own
/// session from the offset each chunk returns.
pub fn list() -> Vec<SessionView> {
    let sessions = SESSIONS.lock();
    sessions
        .iter()
        .filter(|(_, session)| !matches!(session.state, State::Uploading))
        .map(|(id, session)| {
            let written = match session.state {
                State::Placed => session.total,
                _ => disk_len(&staging_path(&session.directory, id)),
            };
            view(id, session, written)
        })
        .collect()
}

fn view(id: &str, session: &Session, written: usize) -> SessionView {
    let detail = match &session.state {
        State::Failed(detail) => Some(detail.clone()),
        _ => None,
    };
    SessionView {
        id: id.to_string(),
        directory: session.directory.clone(),
        name: session.name.clone(),
        path: session.path.clone(),
        total: session.total,
        written,
        state: session.state.name(),
        detail,
    }
}

fn record_failure(id: &str, detail: &str) {
    let mut sessions = SESSIONS.lock();
    if let Some(session) = sessions.get_mut(id) {
        session.state = State::Failed(detail.to_string());
    }
}

fn unknown(id: &str) -> FileError {
    FileError::Unknown(id.to_string())
}

/// Bytes on disk, or zero when the file is not there yet.
fn disk_len(path: &str) -> usize {
    ax_std::fs::metadata(path)
        .map(|metadata| metadata.len() as usize)
        .unwrap_or(0)
}

/// The staging namespace of one target directory.
fn staging_dir(directory: &str) -> String {
    format!("{}/{STAGING_DIR}", directory.trim_end_matches('/'))
}

fn staging_path(directory: &str, id: &str) -> String {
    format!("{}/{id}", staging_dir(directory))
}

/// A session id has to be usable as one path component and nothing else.
fn checked_id(id: &str) -> Result<&str, FileError> {
    let id = id.trim();
    let acceptable = !id.is_empty()
        && id.len() <= SESSION_ID_MAX
        && !id.starts_with('.')
        && !id.contains("..")
        && id
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.' | '+'));
    if acceptable {
        Ok(id)
    } else {
        Err(FileError::InvalidId(id.to_string()))
    }
}

/// The target directory has to be an absolute path inside the guest filesystem.
///
/// `..` is refused rather than resolved: a staging namespace is only meaningful
/// relative to the directory the operator picked, and a path that walks out of
/// it would place files somewhere nobody selected. The staging namespace itself
/// is refused as a target too: bytes are already staged there.
fn checked_directory(directory: &str) -> Result<String, FileError> {
    let trimmed = directory.trim();
    let normalized = if trimmed == "/" {
        "/".to_string()
    } else {
        trimmed.trim_end_matches('/').to_string()
    };
    let acceptable = normalized.starts_with('/')
        && !normalized.contains("..")
        && !normalized.contains("//")
        && !normalized.ends_with(STAGING_DIR);
    if acceptable {
        Ok(normalized)
    } else {
        Err(FileError::InvalidDirectory(directory.to_string()))
    }
}

/// A directory name is one path component: the parent comes from the request.
fn checked_directory_name(name: &str) -> Result<String, FileError> {
    let name = name.trim();
    let acceptable =
        !name.is_empty() && !name.contains('/') && !name.starts_with('.') && !name.contains("..");
    if acceptable {
        Ok(name.to_string())
    } else {
        Err(FileError::InvalidName(name.to_string()))
    }
}

/// A final name is a plain file name: the directory comes from the session.
fn checked_name(name: &str) -> Result<String, FileError> {
    let name = name.trim();
    let acceptable =
        !name.is_empty() && !name.contains('/') && !name.starts_with('.') && !name.contains("..");
    if acceptable {
        Ok(name.to_string())
    } else {
        Err(FileError::InvalidName(name.to_string()))
    }
}

/// Distinguishes "the disk is full" from "the write failed".
///
/// A full disk is the one write failure a client can act on: the offset stays
/// valid, so the upload continues where it stopped once space is freed. The kind
/// is the primary signal; the message is a fallback, because the guest
/// filesystem stack does not guarantee how the out-of-space condition is
/// spelled once it has travelled through it.
fn map_io_error(error: &IoError) -> FileError {
    if *error == IoError::StorageFull {
        FileError::StorageFull(error.to_string())
    } else {
        FileError::Unwritable(error.to_string())
    }
}

/// The same distinction for the calls that report the facade's own error.
///
/// The filesystem operations wrap the io error, so an out-of-space condition
/// arrives one level deeper here than it does from a write.
fn map_std_error(error: &StdError) -> FileError {
    match error {
        StdError::Io(io) => map_io_error(io),
        other => FileError::Unwritable(other.to_string()),
    }
}
